iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp

```diff
@@ -0,0 +1,357 @@
+        }
+    }
+
+    void TearDown()
+    {
+        for (uint32_t i = 0; i < NUMBER_OF_PUBLISHERS; i++)
+        {
+            m_publisherPortUserVector[i].stopOffer();
+            static_cast<void>(m_publisherPortRouDiVector[i].getCaProMessage());
+        }
+
+        m_subscriberPortUserSingleProducer.unsubscribe();
+        m_subscriberPortUserMultiProducer.unsubscribe();
+
+        static_cast<void>(m_subscriberPortRouDiSingleProducer.getCaProMessage());
+        static_cast<void>(m_subscriberPortRouDiMultiProducer.getCaProMessage());
+    }
+
+    GenericRAII m_uniqueRouDiId{[] { iox::popo::internal::setUniqueRouDiId(0); },
+                                [] { iox::popo::internal::unsetUniqueRouDiId(); }};
+
+    uint64_t m_receiveCounter{0};
+    std::atomic<uint64_t> m_sendCounter{0};
+    std::atomic<uint64_t> m_publisherRunFinished{0};
+
+    // Memory objects
+    Allocator m_memoryAllocator{g_memory, MEMORY_SIZE};
+    MePooConfig m_mempoolConfig;
+    MemoryManager m_memoryManager;
+
+    using ConcurrentCaproMessageVector_t = iox::concurrent::smart_lock<vector<CaproMessage, 1>>;
+    ConcurrentCaproMessageVector_t m_concurrentCaproMessageVector;
+    iox::concurrent::smart_lock<vector<CaproMessage, 1>> m_caproMessageRx;
+
+    // subscriber port for single producer
+    SubscriberPortData m_subscriberPortDataSingleProducer{
+        TEST_SERVICE_DESCRIPTION, TEST_SUBSCRIBER_APP_NAME, VariantQueueTypes::SoFi_SingleProducerSingleConsumer};
+    SubscriberPortUser m_subscriberPortUserSingleProducer{&m_subscriberPortDataSingleProducer};
+    SubscriberPortSingleProducer m_subscriberPortRouDiSingleProducer{&m_subscriberPortDataSingleProducer};
+
+    // subscriber port for multi producer
+    SubscriberPortData m_subscriberPortDataMultiProducer{
+        TEST_SERVICE_DESCRIPTION, TEST_SUBSCRIBER_APP_NAME, VariantQueueTypes::SoFi_MultiProducerSingleConsumer};
+    SubscriberPortUser m_subscriberPortUserMultiProducer{&m_subscriberPortDataMultiProducer};
+    SubscriberPortMultiProducer m_subscriberPortRouDiMultiProducer{&m_subscriberPortDataMultiProducer};
+
+    // publisher port
+    vector<PublisherPortData, NUMBER_OF_PUBLISHERS> m_publisherPortDataVector;
+    vector<PublisherPortUser, NUMBER_OF_PUBLISHERS> m_publisherPortUserVector;
+    vector<PublisherPortRouDi, NUMBER_OF_PUBLISHERS> m_publisherPortRouDiVector;
+
+    inline CaproMessage waitForCaproMessage(const ConcurrentCaproMessageVector_t& concurrentCaproMessageVector,
+                                            const CaproMessageType& caproMessageType)
+    {
+        bool finished{false};
+        CaproMessage caproMessage;
+
+        // Wait until the expected CaPro message has arrived in the shared message vector between subscriber and the
+        // first publisher thread
+        do
+        {
+            // Add delay to allow other thread accessing the shared resource
+            std::this_thread::sleep_for(std::chrono::microseconds(10));
+            {
```

Comment:
This is a hard pattern. Maybe two of our cuncurrent FiFos would be more straighforward to exchange the CaPro messages between subscriber and the first publisher. But so far I cannot find a problem with this approach

Reply:
I couldn't make a CaPro message type atomic, hence this approach. I didn't want to mess around with mutex and lock so I used `smart_lock` from iceoryx. Simple handshake between threads. One thread waiting for a specific message before continuing, other thread writing the message. Then vice versa.

Reply:
OK, also not a big issue for the test